pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/RRDD.scala

```diff
@@ -269,26 +269,44 @@ private class StringRRDD[T: ClassTag](
                               packageNames, rLibDir,
                               broadcastVars.map(x => x.asInstanceOf[Broadcast[Object]])) {
 
-  private var dataStream: BufferedReader = _
-
-  override protected def openDataStream(input: InputStream) = {
-    dataStream = new BufferedReader(new InputStreamReader(input))
-    dataStream
-  }
-
-  override protected def read(): String = {
-    try {
-      dataStream.readLine()
-    } catch {
-      case e: IOException => {
-        throw new SparkException("R worker exited unexpectedly (crashed)", e)
-      }
+  override protected def readData(length: Int): String = {
+    length match {
+      case length if length > 0 =>
+        readStringBytes(dataStream, length)
+      case _ => null
     }
   }
 
   lazy val asJavaRDD : JavaRDD[String] = JavaRDD.fromRDD(this)
 }
 
+private object SpecialLengths {
+  val END_OF_STREAM = 0
+  val TIMING_DATA   = -1
+}
+
+private class TimeLogger extends Serializable {
+  var boot: Double = _
+  var init: Double = _
+  var broadcast: Double = _
+  var input: Double = _
+  var compute: Double = _
+  var output: Double = _
+
+  def reportTime: String = {
+    ("Times: boot = %.1f s, init = %.1f s, broadcast = %.1f s, " +
+      "read-input = %.1f s, compute = %.1f s, write-output = %.1f s, " +
+      "total = %.1f s").format(
```

Comment:
This timings are usually a few ms, 1 digit is not enough for them, 3 or 4 will be better.